### PR TITLE
Add drag & drop offline payment settings

### PR DESCRIPTION
**Pull request by ilyasfoo**
### Changes proposed in this Pull Request:

Slack: p1731047871781339/1731031068.064989-slack-C07B44M2B6K

Partially addresses https://github.com/woocommerce/woocommerce/issues/52053:

- Adds `Sortable` component to handle drag and drop
- Adds drag & drop capability to offline payment methods screen

Initially, I tried fixing the issues with existing `Sortable` component from `@woocommerce/components` but I wasn't able to find out the root cause of the component's drag issues. Ultimately, I ended up using a `dndkit` since it's also used by blocks package.

#### Not done in PR

- Add drag & drop capability to main payments screen due to ongoing discussion p6riRB-bfm-p2#comment-11488
- Add backend persistence due to ongoing discussion in above point. Instead, I temporarily use localStorage to establish a basic data structure
 

### How to test the changes in this Pull Request:

1. Enable `reactify-classic-payments-settings` feature flag
2. Go to `WooCommerce > Settings > Payments`
3. Click on `Offline payment gateways`
4. Drag the drag handle and attempt to change ordering
5. Observe you're able to re-order payment method
6. Refresh the screen and observe the ordering is persisted

https://github.com/user-attachments/assets/cfc91302-3ebf-44e6-9d95-fb1b80ee6b2e

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/client/admin/client/settings-payments/hooks/use-payment-ordering.tsx
Comment (ilyasfoo) [inline plugins/woocommerce/client/admin/client/settings-payments/hooks/use-payment-ordering.tsx]: This will eventually need to be replaced with an API call and a backend, spike pending: pe4Hk0-1KO-p2

File: plugins/woocommerce/client/admin/client/settings-payments/hooks/use-payment-ordering.tsx
Comment (ilyasfoo) [inline plugins/woocommerce/client/admin/client/settings-payments/hooks/use-payment-ordering.tsx]: Same as above

File: plugins/woocommerce/client/admin/client/settings-payments/components/offline-payment-gateways/offline-payment-gateways.tsx
Comment (rjchow) [inline plugins/woocommerce/client/admin/client/settings-payments/components/offline-payment-gateways/offline-payment-gateways.tsx]: I'm not sure which PR will land first but it might be easier for us to rebase than moltres - would you consider basing your branch off https://github.com/woocommerce/woocommerce/pull/52824 ?

File: plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx
Comment (rjchow) [inline plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx:33]: :chefskiss

File: plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx
Comment (rjchow) [inline plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx:1]: should we add a todo to move this to @woocommerce/components ?

File: plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx
Comment (rjchow) [inline plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx:50]: Wonder how we might test this SortableContainer in jest tests 🤔 Would dndkit have some reference in their docs or tests?

File: plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx
Comment (rjchow) [inline plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx:53]: I realised I could use the keyboard to manipulate the order thanks to this, by tabbing until the 'enable' button is highlighted then shift-tabbing once, and tapping space followed by arrow keys:
image

It'd be amazing if we could add the focus highlight to add the cherry on top!

File: plugins/woocommerce/client/admin/client/settings-payments/components/offline-gateway-list-item/offline-payment-gateway-list-item.tsx
Comment (rjchow) [inline plugins/woocommerce/client/admin/client/settings-payments/components/offline-gateway-list-item/offline-payment-gateway-list-item.tsx:71]: I've never seen type generics used inline with JSX component tags, TIL!

**Review by rjchow — APPROVED** [review]
This is great! Left some comments

File: plugins/woocommerce/client/admin/client/settings-payments/components/offline-payment-gateways/offline-payment-gateways.tsx
Comment (ilyasfoo) [reply plugins/woocommerce/client/admin/client/settings-payments/components/offline-payment-gateways/offline-payment-gateways.tsx]: You're right. However, it seems a rebase is not ideal at this point since there's some trunk stuff included here that's not included in the other branch and will contain some unrelated code in review. 

I'd opt to merge this after the other branch is merged to trunk.

Comment (rjchow) [reply plugins/woocommerce/client/admin/client/settings-payments/components/offline-payment-gateways/offline-payment-gateways.tsx]: Sure thing!

File: plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx
Comment (ilyasfoo) [reply plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx:1]: Good idea but I think it should be a follow-up issue instead since todos are shunned by lint 😬

Comment (ilyasfoo) [reply plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx:1]: Also, we already have a Sortable components but it's not working 100% and there are still consumers. I wonder if we need to rethink about the name. Either way it's going to be a tech debt 😓

Comment (ilyasfoo) [reply plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx:1]: Created an issue here: https://github.com/woocommerce/woocommerce/issues/52899

File: plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx
Comment (ilyasfoo) [reply plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx:50]: Great idea! I've had help from cursor to generate tests in 31a44ab. However, I don't think we need to test the sensors themselves though, I think that would test the library more than our component.

File: plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx
Comment (ilyasfoo) [reply plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx:53]: Nice find! Tbh I didn't know about this, the sensors were part of a boilerplate 😅 

Anyway, I've added the focus for keyboard in a413cbc

image

File: plugins/woocommerce/client/admin/client/settings-payments/components/offline-gateway-list-item/offline-payment-gateway-list-item.tsx
Comment (ilyasfoo) [reply plugins/woocommerce/client/admin/client/settings-payments/components/offline-gateway-list-item/offline-payment-gateway-list-item.tsx:71]: Yes it's great! It was suggested by cursor 😄

File: plugins/woocommerce/client/admin/client/settings-payments/components/offline-payment-gateways/offline-payment-gateways.tsx
Comment (oaratovskyi) [reply plugins/woocommerce/client/admin/client/settings-payments/components/offline-payment-gateways/offline-payment-gateways.tsx]: > I'd opt to merge this after the other branch is merged to trunk.

@ilyasfoo okay, sounds good, I'll address the PR feedback and merge it!

**Review by adrianduffell — APPROVED** [review]
Tested well, and the code looks reasonable. Good decision to go with localStorage until the backend storage is better understood! 🚀

**Review by psealock — APPROVED** [review]
This is looking good and testing well.

**Review by oaratovskyi — APPROVED** [review]
Looks great! Tested and got acquainted with the code - love the structure and the state context :chefkiss

**Comment by oaratovskyi** [comment]
@rjchow JFYI I've merged the PR with client using wp/data

File: plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx
Comment (rjchow) [reply plugins/woocommerce/client/admin/client/settings-payments/components/sortable/sortable-container.tsx:50]: Oh yea I don't think we need to test the sensors, it was just the trigger for me to think about how testing could be done. Nice work with the tests in [31a44ab](https://github.com/woocommerce/woocommerce/commit/31a44abf83956469e29079fcae06b1e061c98fd3)!

**Comment by ilyasfoo** [comment]
Hey, @oaratovskyi! I've made an [update](https://github.com/woocommerce/woocommerce/pull/52888/commits/484e5213612096372b749528838e1c641038cd98) to accommodate wp.data. Can you take a quick look at that change? 🙏

**Comment by oaratovskyi** [comment]
@ilyasfoo looks great! TIL about introducing https://github.com/woocommerce/woocommerce/pull/52888/commits/484e5213612096372b749528838e1c641038cd98#diff-591d5fdd6e54ce8062870b2a57992f4c3f486958af2c7f740a93ec0c6c51dc81R110 like `localStorage.setItem` in actions where array with type and stuff is returned! thought that it should contain only this 🙈

**Pull request merged**
